packages/arb-token-bridge-ui/src/token-bridge-sdk/Erc20DepositStarter.ts

```diff
@@ -0,0 +1,241 @@
+      address: l2Network.nativeToken,
+      provider: this.sourceChainProvider,
+      owner: address,
+      spender: l1Gateway
+    })
+
+    const gasSummary = await this.transferEstimateGas({ amount, signer })
+    const destinationChainGasPrice =
+      await this.destinationChainProvider.getGasPrice()
+
+    const estimatedDestinationChainGasFeeEth = parseFloat(
+      utils.formatEther(
+        gasSummary.estimatedL2Gas
+          .mul(destinationChainGasPrice)
+          .add(gasSummary.estimatedL2SubmissionCost)
+      )
+    )
+    const estimatedDestinationChainGasFee = utils.parseUnits(
+      String(estimatedDestinationChainGasFeeEth),
+      await nativeCurrency.decimals()
+    )
+
+    // We want to bridge a certain amount of an ERC-20 token, but the Retryable fees on the destination chain will be paid in the custom fee token
+    // We have to check if the native-token spending allowance is enough to cover the fees
+    return customFeeTokenAllowanceForL1Gateway.lt(
+      estimatedDestinationChainGasFee
+    )
+  }
+
+  public async approveNativeCurrency({ signer }: ApproveNativeCurrencyProps) {
+    if (!this.sourceChainErc20Address) {
+      throw Error('Erc20 token address not found')
+    }
+
+    const erc20Bridger = await Erc20Bridger.fromProvider(
+      this.destinationChainProvider
+    )
+    const approveCustomFeeTokenTx = await erc20Bridger.approveFeeToken({
+      erc20L1Address: this.sourceChainErc20Address,
+      l1Signer: signer
+    })
+    await approveCustomFeeTokenTx.wait()
+  }
+
+  public requiresTokenApproval = async ({
+    amount,
+    signer
+  }: RequiresTokenApprovalProps) => {
+    if (!this.sourceChainErc20Address) {
+      throw Error('Erc20 token address not found')
+    }
+
+    const address = await getAddressFromSigner(signer)
+
+    const l1GatewayAddress = await fetchErc20L1GatewayAddress({
+      erc20L1Address: this.sourceChainErc20Address,
+      l1Provider: this.sourceChainProvider,
+      l2Provider: this.destinationChainProvider
+    })
+
+    const allowanceForL1Gateway = await fetchErc20Allowance({
+      address: this.sourceChainErc20Address,
+      provider: this.sourceChainProvider,
+      owner: address,
+      spender: l1GatewayAddress
+    })
+
+    return amount.gte(allowanceForL1Gateway)
+  }
+
+  public async approveTokenEstimateGas({ signer }: ApproveTokenProps) {
+    if (!this.sourceChainErc20Address) {
+      throw Error('Erc20 token address not found')
+    }
+
+    const address = await getAddressFromSigner(signer)
+
+    const l1GatewayAddress = await fetchErc20L1GatewayAddress({
+      erc20L1Address: this.sourceChainErc20Address,
+      l1Provider: this.sourceChainProvider,
+      l2Provider: this.destinationChainProvider
+    })
+
+    const contract = ERC20__factory.connect(
+      this.sourceChainErc20Address,
+      this.sourceChainProvider
+    )
+
+    return contract.estimateGas.approve(
+      l1GatewayAddress,
+      constants.MaxUint256,
```

Comment:
we should allow providing an override amount

Reply:
Updated everywhere